verification/src/block_verifier.rs

```diff
@@ -15,7 +15,11 @@ use rayon::iter::{IndexedParallelIterator, IntoParallelRefIterator, ParallelIter
 use std::collections::HashSet;
 
 //TODO: cellbase, witness
-pub struct BlockVerifier<P> {
+#[derive(Clone)]
+pub struct BlockVerifier<P>
+where
+    P: Clone,
```

Comment:
Is this bound required?

Reply:
The key point is this can be removed 

```
pub struct BlockVerifier<P>
{
    // Verify if the committed transactions is empty
    empty: EmptyVerifier,
    // Verify if the committed and proposed transactions contains duplicate
    duplicate: DuplicateVerifier,
    // Verify the cellbase
    cellbase: CellbaseVerifier<P>,
    // Verify the the committed and proposed transactions merkle root match header's announce
    merkle_root: MerkleRootVerifier,
    // Verify the the uncle
    uncles: UnclesVerifier<P>,
    // Verify the the propose-then-commit consensus rule
    commit: CommitVerifier<P>,
    // Verify all the committed transactions through TransactionVerifier
    transactions: TransactionsVerifier<P>,
}
```

and rewrite 

```
impl<P> BlockVerifier<P>
where
    P: ChainProvider + CellProvider + Clone + 'static,
{
    pub fn new(provider: P) -> Self {
        BlockVerifier {
            // TODO change all new fn's chain to reference
            empty: EmptyVerifier::new(),
            duplicate: DuplicateVerifier::new(),
            cellbase: CellbaseVerifier::new(provider.clone()),
            merkle_root: MerkleRootVerifier::new(),
            uncles: UnclesVerifier::new(provider.clone()),
            commit: CommitVerifier::new(provider.clone()),
            transactions: TransactionsVerifier::new(provider),
        }
    }
}
```

as 

```
impl<P> BlockVerifier<P>
where
    P: Clone,
{
    pub fn new(provider: P) -> Self {
        BlockVerifier {
            // TODO change all new fn's chain to reference
            empty: EmptyVerifier::new(),
            duplicate: DuplicateVerifier::new(),
            cellbase: CellbaseVerifier::new(provider.clone()),
            merkle_root: MerkleRootVerifier::new(),
            uncles: UnclesVerifier::new(provider.clone()),
            commit: CommitVerifier::new(provider.clone()),
            transactions: TransactionsVerifier::new(provider),
        }
    }
}
```
finally we can write as 

```
#[derive(Clone)]
pub struct BlockVerifier<P>
where
    P: Clone,
{
    // Verify if the committed transactions is empty
    empty: EmptyVerifier,
    // Verify if the committed and proposed transactions contains duplicate
    duplicate: DuplicateVerifier,
    // Verify the cellbase
    cellbase: CellbaseVerifier<P>,
    // Verify the the committed and proposed transactions merkle root match header's announce
    merkle_root: MerkleRootVerifier,
    // Verify the the uncle
    uncles: UnclesVerifier<P>,
    // Verify the the propose-then-commit consensus rule
    commit: CommitVerifier<P>,
    // Verify all the committed transactions through TransactionVerifier
    transactions: TransactionsVerifier<P>,
}
```

Reply:
or 

```
#[derive(Clone)]
pub struct BlockVerifier<P>
{
    // Verify if the committed transactions is empty
    empty: EmptyVerifier,
    // Verify if the committed and proposed transactions contains duplicate
    duplicate: DuplicateVerifier,
    // Verify the cellbase
    cellbase: CellbaseVerifier<P>,
    // Verify the the committed and proposed transactions merkle root match header's announce
    merkle_root: MerkleRootVerifier,
    // Verify the the uncle
    uncles: UnclesVerifier<P>,
    // Verify the the propose-then-commit consensus rule
    commit: CommitVerifier<P>,
    // Verify all the committed transactions through TransactionVerifier
    transactions: TransactionsVerifier<P>,
}

#[derive(Clone)]
pub struct BlockVerifier<P>
where
    P: Clone,
{
    // Verify if the committed transactions is empty
    empty: EmptyVerifier,
    // Verify if the committed and proposed transactions contains duplicate
    duplicate: DuplicateVerifier,
    // Verify the cellbase
    cellbase: CellbaseVerifier<P>,
    // Verify the the committed and proposed transactions merkle root match header's announce
    merkle_root: MerkleRootVerifier,
    // Verify the the uncle
    uncles: UnclesVerifier<P>,
    // Verify the the propose-then-commit consensus rule
    commit: CommitVerifier<P>,
    // Verify all the committed transactions through TransactionVerifier
    transactions: TransactionsVerifier<P>,
}
```

I don't think it's a good pattern.